YogaKit/YKLayout.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "YKEnums.h"
+
+@interface YKLayout : NSObject
+
+/**
+ The property that decides if we should include this view when calculating layout. Defaults to YES.
+ */
+@property (nonatomic, setter=setIncluded:) BOOL isIncluded;
+
+/**
+ The property that decides during layout/sizing whether or not yk_* properties should be applied. Defaults to NO.
+ */
+@property (nonatomic, setter=setEnabled:) BOOL isEnabled;
+
+@property (nonatomic) YKDirection direction;
+@property (nonatomic) YKFlexDirection flexDirection;
+@property (nonatomic) YKJustify justifyContent;
+@property (nonatomic) YKAlign alignContent;
+@property (nonatomic) YKAlign alignItems;
+@property (nonatomic) YKAlign alignSelf;
+@property (nonatomic) YKPositionType positionType;
+@property (nonatomic) YKWrap flexWrap;
+
+@property (nonatomic) CGFloat flexGrow;
+@property (nonatomic) CGFloat flexShrink;
+@property (nonatomic) CGFloat flexBasis;
+
+@property (nonatomic) CGFloat positionLeft;
+@property (nonatomic) CGFloat positionTop;
+@property (nonatomic) CGFloat positionRight;
+@property (nonatomic) CGFloat positionBottom;
+@property (nonatomic) CGFloat positionStart;
+@property (nonatomic) CGFloat positionEnd;
+@property (nonatomic) CGFloat positionHorizontal;
```

Comment:
@dshahidehpour The Yoga website says: "We believe that Right-to-Left (RTL) should be a first class citizen when it comes to layout" and I strongly agree. UIKit also supports it by default with leading and trailing constraints in Auto Layout and I'd be very sorry to have them out of YogaKit.

You mention that there are multiple properties that do the same thing. But it's not unheard of in Cocoa. For example, a UIView's `frame.size` is the same as its `bounds.size`. Setting one affects the other. And it's the same with it's `center`, which is the mid-x and mid-y of `frame`.

I know, this is a difficult design decision. But I think there is no perfect solution. For sake of discussion, here are the solutions I see, as well as their pros and cons.

# Properties

```
@property (nonatomic) CGFloat paddingLeft;
@property (nonatomic) CGFloat paddingRight;
@property (nonatomic) CGFloat paddingTop;
@property (nonatomic) CGFloat paddingBottom;
@property (nonatomic) CGFloat paddingStart;
@property (nonatomic) CGFloat paddingEnd;
@property (nonatomic) CGFloat paddingHorizontal;
@property (nonatomic) CGFloat paddingVertical;
@property (nonatomic) CGFloat padding; // I've removed the All because it's a change I'd like to do
```

## Usage in Objective-C

```
view.layout.padding = 15;
// OR
view.layout.paddingTop = 10;
// OR
view.layout.paddingStart = 10;
// OR
view.layout.paddingHorizontal = 10;
view.layout.paddingTop = 15;
```

## Usage in Swift

```
view.layout.padding = 15
// OR
view.layout.paddingTop = 10
// OR
view.layout.paddingStart = 10
// OR
view.layout.paddingHorizontal = 10
view.layout.paddingTop = 10
```

## Pros

* Recognizable API for people from CSS and React Native (it's actually the exact same API as React Native).
* Auto-completion makes this API version fast to use, as for all other properties.

## Cons

* Makes the API surface larger, not very DRY.

# Methods with `NS_ENUM`

```
- (CGFloat)paddingForEdge:(YGEdge)edge;
- (void)setPadding:(CGFloat)padding forEdge:(YGEdge)edge;
```

## Usage in Objective-C

```
[view.layout setPadding:15 forEdge:YGEdgeAll];
// OR
[view.layout setPadding:10 forEdge:YGEdgeTop];
// OR
// NO WAY TO SET START
[view.layout setPadding:10 forEdge:YKEdgeHorizontal];
[view.layout setPadding:10 forEdge:YKEdgeTop];
```

## Usage in Swift

```
view.layout.setPadding(15, for: .all)
// OR
view.layout.setPadding(10, for: .top)
// OR
// NO WAY TO SET START
// OR
view.layout.setPadding(10, for: .horizontal)
view.layout.setPadding(10, for: .top)
```

## Pros

* The usage looks cleaner by having a smaller API surface area.

## Cons

* We can't use dot-notation like all other properties, which breaks the reading flow. Perhaps it's my personal taste, but I see that as a big minus.

```
root.layout.width = 5;
[root.layout setPadding:10 forEdge:YGEdgeTop];
root.layout.flexGrow = 1;
```

# `YGEdgeValue`

```

@interface YGEdgeValue
@property (nonatomic) CGFloat left;
@property (nonatomic) CGFloat right;
@property (nonatomic) CGFloat top;
@property (nonatomic) CGFloat bottom;
@property (nonatomic) CGFloat start;
@property (nonatomic) CGFloat end;
@property (nonatomic) CGFloat horizontal;
@property (nonatomic) CGFloat vertical;
@property (nonatomic) CGFloat all;
@end

@inteface YGLayout
@property (nonatomic, readonly) YGEdgeValue *padding;
@end
```

## Usage in Objective-C

```
view.layout.padding.all = 15;
// OR
view.layout.padding.top = 10;
// OR
view.layout.padding.start = 10;
// OR
view.layout.padding.horizontal = 10;
view.layout.padding.top = 15;
```

## Usage in Swift

```
view.layout.padding.all = 15
// OR
view.layout.padding.top = 10
// OR
view.layout.padding.start = 10
// OR
view.layout.padding.horizontal = 10
view.layout.padding.top = 10
```

## Pros and Cons

This is basically a different flavor of properties. I'm not a hug fan of it because it multiplies the number of objects on the heap that yoga generates by 4 for a small gain in readability and API surface and a small loss in usage typing speed because having to go through two auto-completion hints (`padding`, then `top`).

# Conclusion

On and all, here's my ranking of preference:

1. Properties
2. YGEdgeValue
3. Methods